go/vt/vttablet/tabletserver/messager/message_manager.go

```diff
@@ -541,68 +553,181 @@ func (mm *messageManager) postpone(tsv TabletService, name string, ackWaitTime t
 	}
 }
 
+func (mm *messageManager) startVStream() {
+	mm.streamMu.Lock()
+	defer mm.streamMu.Unlock()
+	if mm.streamCancel != nil {
+		return
+	}
+	var ctx context.Context
+	ctx, mm.streamCancel = context.WithCancel(tabletenv.LocalContext())
+	go mm.runVStream(ctx)
+}
+
+func (mm *messageManager) stopVStream() {
+	mm.streamMu.Lock()
+	defer mm.streamMu.Unlock()
+	if mm.streamCancel != nil {
+		mm.streamCancel()
+		mm.streamCancel = nil
+	}
+}
+
+func (mm *messageManager) runVStream(ctx context.Context) {
+	for {
+		err := mm.runOneVStream(ctx)
+		select {
+		case <-ctx.Done():
+			log.Info("Context canceled, exiting vstream")
+			return
+		default:
+		}
+		MessageStats.Add([]string{mm.name.String(), "VStreamFailed"}, 1)
+		log.Infof("VStream ended: %v, retrying in 5 seconds", err)
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (mm *messageManager) runOneVStream(ctx context.Context) error {
+	var curPos string
+	var fields []*querypb.Field
+
+	err := mm.vs.Stream(ctx, "current", mm.vsFilter, func(events []*binlogdatapb.VEvent) error {
+		mm.streamMu.Lock()
+		defer mm.streamMu.Unlock()
+
+		select {
+		case <-ctx.Done():
+			return io.EOF
+		default:
+		}
+
+		mustSkip := func() (bool, error) {
+			if mm.lastPollPosition == nil {
+				return false, nil
+			}
+			if curPos == "" {
+				return true, nil
+			}
+			cur, err := mysql.DecodePosition(curPos)
+			if err != nil {
+				return false, err
+			}
+			if cur.AtLeast(*mm.lastPollPosition) {
+				mm.lastPollPosition = nil
+				return false, nil
+			}
+			return true, nil
+		}
+		skipEvents, err := mustSkip()
+		if err != nil {
+			return err
+		}
+		var newPos string
+		for _, ev := range events {
+			switch ev.Type {
+			case binlogdatapb.VEventType_FIELD:
+				fields = ev.FieldEvent.Fields
+			case binlogdatapb.VEventType_ROW:
+				if skipEvents {
+					continue
+				}
+				if err := mm.processRowEvent(fields, ev.RowEvent); err != nil {
+					return err
+				}
+			case binlogdatapb.VEventType_GTID:
+				newPos = ev.Gtid
+			case binlogdatapb.VEventType_COMMIT, binlogdatapb.VEventType_DDL, binlogdatapb.VEventType_OTHER:
+				curPos = newPos
+				skipEvents, err = mustSkip()
+				if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+	return err
+}
+
+func (mm *messageManager) processRowEvent(fields []*querypb.Field, rowEvent *binlogdatapb.RowEvent) error {
+	if fields == nil {
+		// Unreachable.
+		return fmt.Errorf("internal error: unexpected rows without fields")
+	}
+
+	now := time.Now().UnixNano()
+	for _, rc := range rowEvent.RowChanges {
+		if rc.After == nil {
+			continue
+		}
+		row := sqltypes.MakeRowTrusted(fields, rc.After)
+		mr, err := BuildMessageRow(row)
+		if err != nil {
+			return err
+		}
+		if mr.TimeNext == 0 || mr.TimeNext > now {
```

Comment:
I think this changes the current semantics, though it might be a good change. Our devs often will set `time_next=0` to requeue messages during debugging. I believe now that causes an automatic add to the cache. Here, the special case of `time_next=0` doesn't add to the cache, and will instead be picked up by the poller. Is there a use case you can think of where that behavior should be intentional? If there is a good use case, I would just train devs to set `time_next=1` while debugging, and we can add this to the docs.

Reply:
Sounds good to me. We can update the docs to warn against using `time_next=0`